### PR TITLE
shared_ptr: ignore false alarms from GCC-12

### PR DESCRIPTION
**Pull request by tchaikov**
before GCC addresses this issue, we have to disable this warning.

see also https://github.com/scylladb/seastar/issues/1037 and https://gcc.gnu.org/bugzilla/show_bug.cgi?id=105204.

Signed-off-by: Kefu Chai <tchaikov@gmail.com>

### CODE REVIEW

**Comment by tchaikov** [comment]
being tested at https://github.com/tchaikov/seastar/pull/1

**Comment by tchaikov** [comment]
```
63/74 Test #63: Seastar.unit.tls ..............................***Failed   46.88 sec
[0/1] cd /home/circleci/project/build/release && /home/circleci/project/build/release/tests/unit/tls_test -- -c 2
Running 29 test cases...
INFO  2022-11-30 05:49:22,691 seastar - Reactor backend: linux-aio
WARN  2022-11-30 05:49:22,691 [shard 0] seastar - Creation of perf_event based stall detector failed, falling back to posix timer: std::system_error (error system:13, perf_event_open() failed: Permission denied)
INFO  2022-11-30 05:49:22,692 [shard 0] seastar - Created fair group io-queue-0, capacity rate 2147483:2147483, limit 12582912, rate 16777216 (factor 1), threshold 2000
INFO  2022-11-30 05:49:22,692 [shard 0] seastar - IO queue uses 0.75ms latency goal for device 0
INFO  2022-11-30 05:49:22,692 [shard 0] seastar - Created io group dev(0), length limit 4194304:4194304, rate 2147483647:2147483647
INFO  2022-11-30 05:49:22,692 [shard 0] seastar - Created io queue dev(0) capacities: 512:2000:2000 1024:3000:3000 2048:5000:5000 4096:9000:9000 8192:17000:17000 16384:33000:33000 32768:65000:65000 65536:129000:129000 131072:257000:257000
WARN  2022-11-30 05:49:22,705 [shard 1] seastar - Creation of perf_event based stall detector failed, falling back to posix timer: std::system_error (error system:13, perf_event_open() failed: Permission denied)
random-seed=435077401
Established connection
Killing server side
Got expected exception
terminate called after throwing an instance of 'std::filesystem::__cxx11::filesystem_error'
  what():  filesystem error: open failed: No such file or directory [/tmp/test8e4c/..]
Aborting on shard 0.
Backtrace:
  0x4b90da
  0x4eee42
  /lib/x86_64-linux-gnu/libc.so.6+0x4251f
  /lib/x86_64-linux-gnu/libc.so.6+0x96a7b
  /lib/x86_64-linux-gnu/libc.so.6+0x42475
  /lib/x86_64-linux-gnu/libc.so.6+0x287f2
  /lib/x86_64-linux-gnu/libstdc++.so.6+0xa2bfd
  /lib/x86_64-linux-gnu/libstdc++.so.6+0xae28b
  /lib/x86_64-linux-gnu/libstdc++.so.6+0xae2f6
  0x44601a
  0x447412
  0x42e67f
  0x45c03c
  0x54c671
Aborted (core dumped)
FAILED: tests/unit/CMakeFiles/test_unit_tls_run /home/circleci/project/build/release/tests/unit/CMakeFiles/test_unit_tls_run 
cd /home/circleci/project/build/release && /home/circleci/project/build/release/tests/unit/tls_test -- -c 2
ninja: build stopped: subcommand failed.
```

the test failure is not relevant to this change.

File: include/seastar/core/shared_ptr.hh [hunk 1]
```diff
@@ -290,17 +290,31 @@ public:
     lw_shared_ptr(std::nullptr_t) noexcept : lw_shared_ptr() {}
     lw_shared_ptr(const lw_shared_ptr& x) noexcept : _p(x._p) {
         if (_p) {
+#pragma GCC diagnostic push
+        // to silence the false alarm from GCC 12, see
+        // https://gcc.gnu.org/bugzilla/show_bug.cgi?id=105204
+#if defined(__GNUC__) && (__GNUC__ >= 12)
```
Comment (avikivity) [inline include/seastar/core/shared_ptr.hh:296]: https://godbolt.org/z/dqY4xrTKT

Comment (tchaikov) [reply include/seastar/core/shared_ptr.hh]: ahh, thanks for pointing this out. forgot that clang "pretended" to be GCC!

**Pull request merged**